notifications_utils/decorators.py

```diff
@@ -14,3 +20,116 @@ def wrapper(*args, **kwargs):
+
+    # Ensure chunk size remains within min and max chunk size bounds
+    chunk_size = max(MIN_CHUNK_SIZE, min(chunk_size, MAX_CHUNK_SIZE))
+    worker_count = math.ceil(data_size / chunk_size)
+
+    # Suppress workers for larger chunks to avoid memory and/or context switching overhead
+    if chunk_size > MAX_CHUNK_SIZE * 0.8:
+        worker_count = min(worker_count, MAX_WORKERS // 2)
+    else:
+        worker_count = min(worker_count, MAX_WORKERS)
+
+    return chunk_size, worker_count
+
+
+# Parallel processing decorator
+def parallel_process_iterable(chunk_size=10000, max_workers=None, break_condition: Optional[Callable] = None):
+    """Decorator to split processing an iterable into chunks and execute in parallel. This should decorate the function responsible for processing each chunk.
+    If processing can be stopped early, this condition should be defined in the processing function, and the `break_condition` parameter should be provided.
+
+    `chunk_size` and `max_workers` are managed internally to optimize performance based on the size of the data to be processed, but can be overridden if necessary.
+
+    Args:
+        chunk_size (int, optional): Defaults to 10,000
+        max_workers (_type_, optional): Defaults to the number of CPU cores
+        break_condition (function, optional): A lambda function that defines when parallel execution can be stopped. This applies to the entire iterable, not just the
+        current chunk. When any of the threads returns a result that satisfies the break condition, the processing is stopped and the results are returned.
+    """
+
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            data = args[0]
+            data_size = len(data)
+            nonlocal chunk_size, max_workers
+            chunk_size, max_workers = control_chunk_and_worker_size(data_size, chunk_size, max_workers)
+
+            def process_chunk(chunk):
+                return func(chunk, *args[1:])
+
+            # Execute in parallel
+            with ThreadPoolExecutor(max_workers=max_workers) as executor:
+                futures = [executor.submit(process_chunk, chunk) for chunk in chunk_iterable(data, chunk_size)]
+                current_app.logger.info(
+                    f"Beginning parallel processing of {data_size} items in {len(futures)} chunks for {func.__name__} across {max_workers} workers."
+                )
+                # Combine results
+                results = []
+                for future in futures:
+                    result = future.result()
+                    results.append(result)
+                    if break_condition:
+                        try:
+                            if break_condition(result):
+                                return results
+                        except Exception as e:
+                            current_app.logger.warning(f"Break condition exception: {e}")
```

Comment:
Are the exceptions swallowed in the parallelization if we don't log these?

Reply:
They are not swallowed and will raise if thrown in a thread. I think there is some value in catching and identifying that they were caused by one of the running threads / the break condition.

This also raises another question, in the except we can choose to raise it and stop processing, or log it, continue, and allow the remaining threads to finish processing. It seems like a choice that depends on the context. Perhaps an `is_atomic` flag could allow control over such behaviour?